readme.txt
=== Mobio for WooCommerce ===
Tags: woocommerce, payments, alternative payments, sms payments
Requires at least: 4.7
Tested up to: 6.0
Requires PHP: 7.2
Stable tag: 1.0.0
License: Apache 2.0

== Description ==

Mobio for WooCommerce is secure solution for alternative SMS payment with WooCommerce.

=== Benefits and possibilities of using Mobio for woocommerce plugin ===

=== Supported industries ===

== Instalation ==

=== Install from within WordPress ===

1. Visit the plugins page within your dashboard and select ‘Add New’;
2. Search for ‘Mobio for wooCommerce’;
3. Activate ‘Mobio for wooCommerce from your Plugins page;
4. Go to ‘after activation’ below.

=== Install manually ===

1. Upload the ‘mobio-for-woocommerce’ folder to the /wp-content/plugins/ directory;
2. Activate the ‘Mobio for wooCommerce plugin through the ‘Plugins’ menu in WordPress;
3. Go to ‘after activation’ below.

=== After activation ===

== Frequently Asked Questions ==

=== What is Mobio for WooCommerce?
Mobio for WooCommerce is a Payment Gateway that allows you to accept payments with SMS.

=== Is Mobio account required?
Yes! A Mobio account is required to receive the necessary credentials for Mobio.

=== How to get started?
Download, install and activate Mobio for WooCommerce from Plugins page in WordPress administration.

== Screenshots ==

== Changelog ==

See changelog.txt


changelog.txt
*** Mobio for WooCommerce Changelog ***

2022-05-21 - version 1.0.0
* Initial release


functions.php
<?php
/**
 * Main functions
 *
 * @package MobioWooCommerce
 * @author  Zota
 */

if ( ! defined( 'ABSPATH' ) ) {
	exit;
}


/**
 * Check if requirements are ok.
 *
 * @return bool
 */
function wc_gateway_mobio_requirements() {
	// Check if all requirements are ok.
	$woocommerce_active  = class_exists( 'woocommerce' );
	$woocommerce_version = defined( 'WC_VERSION' ) && version_compare( WC_VERSION, MOBIO_WC_MIN_WC_VER, '>=' );
	$php_version         = version_compare( PHP_VERSION, MOBIO_WC_MIN_PHP_VER, '>=' );

	return $woocommerce_active && $woocommerce_version && $php_version;
}


/**
 * WooCommerce not activated error message.
 *
 * @return void
 */
function wc_gateway_mobio_woocommerce_error() {
	global $pagenow;

	if ( 'plugins.php' !== $pagenow ) {
		return;
	}

	?>
	<div class="notice notice-warning">
		<p>
			<?php
			printf(
				// translators: %1$s is plugin name.
				esc_html__( '%1$s requires WooCommerce to be active.', 'mobio-woocommerce' ),
				'<strong>' . esc_html( MOBIO_WC_NAME ) . '</strong>'
			);
			?>
			</strong>
		</p>
	</div>
	<?php
}


/**
 * Plugin initialization.
 *
 * @return void
 */
function mobio_plugin_init() {
	// Load the textdomain.
	load_plugin_textdomain( 'mobio-woocommerce', false, plugin_basename( dirname( __FILE__, 2 ) ) . '/languages' );

	// Check requirements
	if ( ! wc_gateway_mobio_requirements() ) {
		add_action( 'admin_notices', 'wc_gateway_mobio_requirements_error' );
		return;
	}

	// Show admin notice if WooCommerce is not active.
	if ( ! class_exists( 'woocommerce' ) ) {
		add_action( 'admin_notices', 'wc_gateway_mobio_woocommerce_error' );
		return;
	}

	wc_gateway_mobio_init();
}


/**
 * Gateway init.
 *
 * @return void
 */
function wc_gateway_mobio_init() {
	// Register admin scripts.
	add_action( 'admin_enqueue_scripts', 'mobio_admin_enqueue_scripts' );

	// Register scripts.
	add_action( 'wp_enqueue_scripts', 'mobio_enqueue_scripts' );

	// Hook filters and actions.
	add_filter( 'woocommerce_gateway_description', array( 'Mobio_WooCommerce', 'checkout_form_field' ), 20, 2 );
	add_action( 'woocommerce_product_options_general_product_data', array( '\Unax\Mobio\Includes\Settings', 'product_options_general_product_data' ) );
	add_action( 'woocommerce_process_product_meta', array( '\Unax\Mobio\Includes\Settings', 'process_product_meta' ) );

	// Initialize.
	require_once MOBIO_WC_PATH . '/includes/class-mobio-woocommerce.php';

	// Add to woocommerce payment gateways.
	add_filter(
		'woocommerce_payment_gateways',
		function ( $gateways ) {
			$gateways[] = new Mobio_WooCommerce();
			return $gateways;
		}
	);

	// add_filter( 'woocommerce_gateway_description', array(  ), 20, 2 );
}


/**
 * Admin options scripts.
 *
 * @param  string $hook WooCommerce Hook.
 * @return void
 */
function mobio_admin_enqueue_scripts( $hook ) {
	if ( 'woocommerce_page_wc-settings' !== $hook ) {
		return;
	}

	wp_enqueue_style( 'mobio-woocommerce-admin', MOBIO_WC_URL . 'dist/css/admin.css', array(), MOBIO_WC_VERSION );

	wp_enqueue_script( 'mobio-polyfill', MOBIO_WC_URL . 'dist/js/polyfill.js', array(), MOBIO_WC_VERSION, true );
	wp_enqueue_script( 'mobio-woocommerce', MOBIO_WC_URL . 'dist/js/admin.js', array( 'wp-i18n', 'jquery', 'mobio-polyfill' ), MOBIO_WC_VERSION, true );
}


/**
 * Public scripts.
 *
 * @return void
 */
function mobio_enqueue_scripts() {
	if ( is_admin() ) {
		return;
	}

	wp_register_style( 'mobio-woocommerce', MOBIO_WC_URL . 'dist/css/styles.css', array( 'woocommerce-inline' ), MOBIO_WC_VERSION );

	if ( is_checkout() ) {
		wp_enqueue_style( 'mobio-woocommerce' );
	}
}


/**
 * Deactivate plugin.
 *
 * @return void
 */
function wc_gateway_mobio_deactivate() {}


includes/class-settings.php
<?php
/**
 * Mobio for WooCommerce Settings.
 *
 * @package MobioWooCommerce
 * @author  Unax
 */

namespace Unax\Mobio\Includes;

if ( ! defined( 'ABSPATH' ) ) {
	exit;
}

/**
 * Settings class.
 */
class Settings {

	/**
	 * Test mode
	 *
	 * @var string
	 */
	public static $test_mode;

	/**
	 * Admin
	 */
	public static function form_fields() {
		$woocommerce_currency = get_woocommerce_currency();

		// @codingStandardsIgnoreStart
		return apply_filters(
			'wc_mobio_settings',
			array(
				'enabled'                     => array(
					'title'   => esc_html__( 'Enable/Disable', 'mobio-woocommerce' ),
					'label'   => sprintf( '%s %s', esc_html__( 'Enable', 'mobio-woocommerce' ), MOBIO_WC_NAME ),
					'type'    => 'checkbox',
					'default' => 'yes',
				),
				'title'                       => array(
					'title'       => esc_html__( 'Title', 'mobio-woocommerce' ),
					'description' => esc_html__( 'This controls the title which the user sees during checkout.', 'mobio-woocommerce' ),
					'default'     => esc_html__( 'SMS Code', 'mobio-woocommerce' ),
					'type'        => 'text',
					'desc_tip'    => true,
				),
				'test_mode'                    => array(
					'title'   => esc_html__( 'Test Mode', 'mobio-woocommerce' ),
					'label'   => esc_html__( 'Enable Test Mode', 'mobio-woocommerce' ),
					'type'    => 'checkbox',
					'default' => 'yes',
				),
			)
		);
		// @codingStandardsIgnoreEnd
	}


	/**
	 * Get payment method description.
	 *
	 * @return string
	 */
	public static function get_description() {

		return wp_kses(
			sprintf(
				// translators: %1$s Plugin name, %2$s Mobio registration url, %3$s Mobio in english.
				__(
					'%1$s provides payments by SMS. <a href="%2$s" target="_blank">Sign up</a> for %1$s account.',
					'mobio-woocommerce'
				),
				'Mobio',
				'https://mobio.bg/site/registration'
			),
			array(
				'a'     => array(
					'href'   => array(),
					'target' => array(),
				),
				'br'    => array(),
				'small' => array(),
			)
		);

	}


	/**
     * Add the fields to product general tab in admin.
     */
    public static function product_options_general_product_data() {
        global $post;

		$product = wc_get_product( $post->ID );

        woocommerce_wp_text_input(
			array(
	            'id'            => 'mobio_service_id',
	            'label'         => esc_html__( 'Mobio service ID', 'mobio-woocommerce' ),
	            'description'   => '',
	            'value'         => $product->get_meta( '_mobio_service_id', true ),
	        )
		);

		woocommerce_wp_text_input(
			array(
	            'id'            => 'mobio_number',
	            'label'         => esc_html__( 'Mobio number', 'mobio-woocommerce' ),
	            'description'   => '',
	            'value'         => $product->get_meta( '_mobio_number', true ),
	        )
		);

		woocommerce_wp_text_input(
			array(
	            'id'            => 'mobio_text',
	            'label'         => esc_html__( 'Mobio text', 'mobio-woocommerce' ),
	            'description'   => '',
	            'value'         => $product->get_meta( '_mobio_text', true ),
	        )
		);
    }


    /**
     * Save product meta.
	 *
	 * @param int $post_id Post ID.
     */
    public static function process_product_meta( $post_id ) {
        $product = wc_get_product( $post_id );

		if ( isset( $_POST['mobio_service_id'] ) ) {
			$product->update_meta_data( '_mobio_service_id', (int) $_POST['mobio_service_id'] );
		}

		if ( isset( $_POST['mobio_number'] ) ) {
			$product->update_meta_data( '_mobio_number', (int) $_POST['mobio_number'] );
		}

		if ( isset( $_POST['mobio_text'] ) ) {
			$product->update_meta_data( '_mobio_text', sanitize_text_field( $_POST['mobio_text'] ) );
		}

        $product->save();
    }
}


includes/class-mobio-woocommerce.php
<?php
/**
 * Mobio for WooCommerce
 *
 * @package MobioWooCommerce
 * @author  Unax
 */

use Unax\Mobio\Includes\Settings;

if ( ! defined( 'ABSPATH' ) ) {
	exit;
}

/**
 * Mobio_WooCommerce class.
 *
 * @extends WC_Payment_Gateway
 */
class Mobio_WooCommerce extends \WC_Payment_Gateway {

	/**
	 * Defines main properties, load settings fields and hooks
	 */
	public function __construct() {

		$this->id                 = MOBIO_WC_GATEWAY_ID;
		$this->method_title       = MOBIO_WC_NAME;
		$this->method_description = Settings::get_description();
		$this->has_fields         = false;
		$this->supports           = array(
			'products',
		);

		// Load the form fields.
		$this->init_form_fields();

		// Load the settings.
		$this->init_settings();

		// Define user set variables.
		$this->title          = $this->get_option( 'title' );
		$this->description    = $this->get_option( 'description' );
		$this->payment_method = $this->get_option( 'payment_method' );

		// Hooks.
		add_action( 'woocommerce_update_options_payment_gateways_' . $this->id, array( $this, 'process_admin_options' ) );
	}


	/**
	 * Check if the currency is in the supported currencies
	 *
	 * @return bool
	 */
	public function is_supported() {
		return true;
	}


	/**
	 * Check if the gateway is available.
	 *
	 * @return bool
	 */
	public function is_available() {
		if ( ! $this->is_supported() ) {
			return false;
		}

		return parent::is_available();
	}


	/**
	 * Initialise Gateway Settings Form Fields
	 */
	public function init_form_fields() {
		$this->form_fields = Settings::form_fields();
	}


	/**
	 * Process the payment and return the result
	 *
	 * @param string $description Payment gateway description.
	 * @param string $payment_id  Payment gateway ID.
	 *
	 * @return string
	 */
	public static function checkout_form_field( $description, $payment_id ) {
		if( MOBIO_WC_GATEWAY_ID !== $payment_id ) {
			return;
		}

		if ( WC()->cart->is_empty() ) {
			return;
		}

		$mobio_service_id = '';
		$mobio_number     = '';
		$mobio_text       = '';
		$mobio_price      = '';

		// Loop over $cart items
		foreach ( WC()->cart->get_cart() as $cart_item_key => $cart_item ) {
			$product = $cart_item['data'];

			$mobio_service_id = $product->get_meta( '_mobio_service_id', true );
			$mobio_number     = $product->get_meta( '_mobio_number', true );
			$mobio_text       = $product->get_meta( '_mobio_text', true );
			$mobio_price      = $product->is_on_sale() ? $product->get_sale_price() : $product->get_regular_price();
		}

		$description = wp_kses(
			sprintf(
				// translators: %s SMS text, %s Mobile number, %s Price.
				__(
					'Please send SMS with text <strong>%s</strong> to <strong>%s</strong> and place the received code here. Price: %s',
					'mobio-woocommerce'
				),
				esc_html( $mobio_text ),
				(int) $mobio_number,
				wc_price( $mobio_price )
			),
			array(
				'strong'    => array(),
			)
		);

		ob_start(); // Start buffering

		printf( '<input type="hidden" name="mobio_service_id" value="%s">', (int) $mobio_service_id );

		woocommerce_form_field(
			'mobio_code',
			array(
				'type'          => 'text',
				'label'         => '',
				'class'         => array( 'form-row-wide' ),
				'required'      => false,
			),
			''
		);
		$description .= ob_get_clean();

	    return $description;
	}


	/**
	 * Process the payment and return the result
	 *
	 * @param  int $order_id Order ID.
	 * @return array
	 */
	public function process_payment( $order_id ) {
		$order = wc_get_order( $order_id );
		if ( empty( $order ) ) {
			wc_add_notice(
				esc_html__( 'Order not found.', 'mobio-woocommerce' ),
				'error'
			);
			return false;
		}

		// Prepare fields.
		$mobio_service_id = isset( $_POST['mobio_service_id'] ) ? (int) $_POST['mobio_service_id'] : null;
		$mobio_code       = isset( $_POST['mobio_code'] ) ? sanitize_text_field( $_POST['mobio_code'] ) : null;

		// Check fields.
		if ( empty( $mobio_service_id ) ) {
			wc_add_notice( esc_html__( 'Bad request: Mobio Service ID missing', 'mobio-woocommerce' ), 'error' );
			return false;
		}

		if ( empty( $mobio_code ) ) {
			wc_add_notice( sprintf( esc_html__( 'Missing required field for the selected payment method %s', 'mobio-woocommerce' ), $this->title ), 'error' );
			return false;
		}

		// Code validation request.
		$args  = array(
			'servID' => $mobio_service_id,
			'code' => $mobio_code
		);

		// Is test mode.
		$settings     = get_option( 'woocommerce_' . MOBIO_WC_GATEWAY_ID . '_settings', array() );
		$is_test_mode = ! empty( $settings['test_mode'] ) && 'yes' === $settings['test_mode'];
		if ( $is_test_mode ) {
			$args  = array(
				'servID' => 29,
				'code' => 'T4JC6G'
			);
		}

		// Build URL.
		$url = sprintf( 'http://www.mobio.bg/code/checkcode.php?%s', http_build_query( $args ) );

		$response = wp_remote_get( $url );
		if ( is_wp_error( $response ) ) {
			return false;
		}

		// Check if is successful payment.
		if ( 'PAYBG=OK' !== $response['body'] ) {
			return false;
		}

		// Process order.
		$order = wc_get_order( $order_id );

		$order->payment_complete();
		$order->update_status( 'completed', __( 'Order paid with SMS', 'mobio-woocommerce' ) );

		// Empty cart.
		WC()->cart->empty_cart();

		return array(
			'result'   => 'success',
			'redirect' => $this->get_return_url( $order )
		);
	}


	/**
	 * Get payment method icon.
	 *
	 * @return string
	 */
	public function get_icon() {
		$attachment = null;
		if ( ! empty( $this->get_option( 'icon' ) ) ) {
			$atts       = array(
				'class' => 'mobio-icon gateway-' . hash( 'crc32b', $this->id ) . '-icon',
			);
			$attachment = wp_get_attachment_image( $this->get_option( 'icon' ), 'medium', false, $atts );
		}

		$icon = apply_filters( 'mobio_woocommerce_' . $this->id . '_icon', $attachment );
		if ( empty( $icon ) ) {
			return;
		}

		return $icon;
	}
}


mobio-for-woocommerce.php
<?php
/**
 * Plugin Name: Mobio for WooCommerce
 * Description: A plugin provides payment gateway for SMS payment with Mobio.
 * Version: 1.0.0
 * Requires at least: 4.7
 * Requires PHP: 7.2
 * Author: Unax
 * Author URI: https://unax.org/
 * Text Domain: mobio-woocommerce
 *
 * WC requires at least: 3.0
 * WC tested up to:  6.4.1
 *
 * License: Apache-2.0
 * License URI: https://github.com/mobio/mobio-woocommerce/blob/master/LICENSE
 *
 * @package     MobioWooCommerce
 * @author      Unax
 */

if ( ! defined( 'ABSPATH' ) ) {
	exit; // Exit if accessed directly.
}

// Set constants.
define( 'MOBIO_WC_NAME', 'Mobio for WooCommerce' );
// Note: this gets replaced at runtime by Github Actions during release, keep the version '1.1.1'.
define( 'MOBIO_WC_VERSION', '1.0.0' );
define( 'MOBIO_WC_GATEWAY_ID', 'wc_mobio' );
define( 'MOBIO_WC_PLUGIN_ID', 'mobio' );
define( 'MOBIO_WC_MIN_PHP_VER', '7.2.0' );
define( 'MOBIO_WC_MIN_WC_VER', '3.0' );
define( 'MOBIO_WC_PATH', plugin_dir_path( __FILE__ ) );
define( 'MOBIO_WC_URL', plugin_dir_url( __FILE__ ) );

// Includes.
require_once MOBIO_WC_PATH . 'functions.php';
require_once MOBIO_WC_PATH . 'includes/class-settings.php';

// Init.
add_action( 'plugins_loaded', 'mobio_plugin_init' );

/**
 * Register deactivation hook.
 */
register_deactivation_hook( __FILE__, 'wc_gateway_mobio_deactivate' );
